src/Network/MQTT/Client.hs

```diff
@@ -18,31 +20,40 @@ import           Data.ByteString.Builder       ( hPutBuilder )
 import           Network.Attoparsec            ( parseMany )
 import           Control.Concurrent.STM.TQueue ( TQueue, newTQueue, readTQueue
                                                , writeTQueue )
-import           Control.Concurrent.STM.TVar   ( TVar, newTVar, writeTVar )
-import           Control.Monad                 ( forM_ )
+import           Control.Concurrent.STM.TVar   ( TVar, newTVar, writeTVar, modifyTVar, readTVar )
+import           Control.Monad                 ( forM_, when )
 import           Control.Monad.STM             ( STM, atomically )
 import           Control.Concurrent            ( forkIO )
+import           Data.List                     ( partition, filter )
+import qualified Data.Map.Strict               as Map
+import           Data.Maybe                    ( isJust, fromJust )
 
 type SocketHandle = (Socket, Handle)
 
 data Config = Config { hostName :: HostName
                      , port     :: String
                      }
 
+data Subscription = Subscription { maxQoS :: QoS -- ^ maximum QoS that was granted for that Subscription
+                                 } deriving (Eq)
+
 data ClientState =
   ClientState
-  { sendQueue        :: TQueue Packet -- ^ All packets from the this queue will be sent to the Server
-  , connected        :: TVar Bool -- ^ Whether the Client is connected to the Server
-  , sessionPresent   :: TVar Bool -- ^ Whether the Server has Session for the Client
-  , pingRespReceived :: TVar Bool -- ^ Whether the Client received Pingresp packet from the Server
-  }
-  deriving (Eq)
+  { sendQueue                 :: TQueue Packet -- ^ All packets from the this queue will be sent to the Server
+  , connected                 :: TVar Bool -- ^ Whether the Client is connected to the Server
+  , sessionPresent            :: TVar Bool -- ^ Whether the Server has Session for the Client
+  , pingRespReceived          :: TVar Bool -- ^ Whether the Client received Pingresp packet from the Server
+  , notAcknSubscribePackets   :: TVar [SubscribePacket] -- ^ Keep Subscribe packets which have not been acknowledged
```

Comment:
`notAckSubscribePackets`. I meant `Acknowledgement` is commonly shortened as `Ack`, not `Ackn`.